### PR TITLE
eth/filters, ethclient, ethereum: API improvements

### LINKED ISSUE #3331: ethclient: "invalid memory address or nil pointer dereference" in BlockByNumber
I have a program running that regularly checks for new blocks and looks for interesting transactions in it and for some reason at some points it panics.

Tell me if I can provide more information. I'm on commit 2c802399c3ebae86af6609c5c0a13058c1017810
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x6371b7]

goroutine 23 [running]:
panic(0x81d960, 0xc42000c120)
	/usr/lib/go/src/runtime/panic.go:500 +0x1a1
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types.CopyHeader(0x0, 0xc4203ae380)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types/block.go:315 +0x57
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types.(*Block).WithBody(0xc4202d5320, 0xc420402a80, 0x1c, 0x1c, 0xc42044c290, 0x2, 0x2, 0x0)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types/block.go:453 +0x1f9
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient.(*Client).getBlock(0xc4200f00b0, 0x7f415bcf1128, 0xc42000c608, 0x8a0aec, 0x14, 0xc4205a9560, 0x2, 0x2, 0xc420563d80, 0xc420563d80, ...)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient/ethclient.go:127 +0x767
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient.(*Client).BlockByNumber(0xc4200f00b0, 0x7f415bcf1128, 0xc42000c608, 0x0, 0xc42000c608, 0x0, 0x0)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient/ethclient.go:69 +0x14e
main.(*TokenNotifier).run(0xc420202ae0, 0x28e411)
	/home/steven/gocode/src/github.com/san-lab/gastracker/tokennotifier.go:65 +0x28d
created by main.StartTokenNotifier
	/home/steven/gocode/src/github.com/san-lab/gastracker/tokennotifier.go:47 +0x2b2
```
```
2016/11/23 15:50:12 No new blocks, sleeping...
2016/11/23 15:50:14 No new blocks, sleeping...
2016/11/23 15:50:16 No new blocks, sleeping...
2016/11/23 15:50:18 No new blocks, sleeping...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x62d187]

goroutine 14 [running]:
panic(0x813080, 0xc42000c120)
	/usr/lib/go/src/runtime/panic.go:500 +0x1a1
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types.CopyHeader(0x0, 0xc420030a18)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types/block.go:315 +0x57
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types.(*Block).WithBody(0xc42177c1b0, 0xc4203d3020, 0x1, 0x4, 0xc4203eefa0, 0x2, 0x2, 0x0)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/core/types/block.go:453 +0x1f9
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient.(*Client).getBlock(0xc4200302a8, 0x7f9df7b16168, 0xc42000c608, 0x895623, 0x14, 0xc4203baa00, 0x2, 0x2, 0xc4218bbd78, 0xc4218bbd78, ...)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient/ethclient.go:127 +0x767
github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient.(*Client).BlockByNumber(0xc4200302a8, 0x7f9df7b16168, 0xc42000c608, 0x0, 0xc42000c608, 0x0, 0x0)
	/home/steven/gocode/src/github.com/san-lab/gastracker/vendor/github.com/ethereum/go-ethereum/ethclient/ethclient.go:69 +0x14e
main.(*TokenNotifier).run(0xc4201c57a0, 0x28e6ed)
	/home/steven/gocode/src/github.com/san-lab/gastracker/tokennotifier.go:65 +0x28d
created by main.StartTokenNotifier
	/home/steven/gocode/src/github.com/san-lab/gastracker/tokennotifier.go:47 +0x2b2
```

### LINKED ISSUE #3379: RFC: Go API handling of missing items
This is a design question about the top-level Go API. What should API methods return if an
item (block, header, transaction, receipt) does not exist? Package ethclient returns `nil`
for the item and `nil` as the error, mirroring the RPC API.

The decision here will make its way into docs, examples, etc.

### To err or not to err

```go
head, err := client.BlockByHash(ctx, common.Hash{})
// head and err are nil
```

This looks fine at first because the non-existence of an item is a special condition that
isn't necessarily an error. But it's also easy to forget and can lead to crash-prone code
because each call must check for a nil result before using it.

An alternative solution to this is to return a non-nil error in case of non-existence.
Callers already assume that the result is not useable if the error is non-nil, which would
protect against accidental crashes where the check is missing. But it's still useful to be
able to check for this specific error (similar to `io.EOF`).

### What should the error look like

Assuming that we want to add an error for non-existence, these are our options.
 Since the API is defined as a set of interfaces and users of the API should avoid depending
on a specific implementation, the error must be defined in the top-level ethereum package.

#### ethereum.NotFound

```go
package ethereum

var NotFound = errors.New("not found")
```

This is the simplest way, mirroring `io.EOF` directly. It's easy to use:

```go
head, err := client.BlockByHash(ctx, common.Hash{})
if err != nil && err != ethereum.NotFound {
     ... 
}
// head might still be nil, but the check is explicit.
```

#### NotFound method on error

```go
package ethereum

type NotFound interface{
    NotFound() bool
}
```

This error is more cumbersome to check for, but allows placing contextual information
into the error (i.e. the message can be different depending on the implementation and
type of item).

```go
head, err := client.BlockByHash(ctx, common.Hash{})
if enf, ok := err.(ethereum.NotFound); ok && enf.NotFound() {
     // now you know
} else if err != nil {
     fmt.Println("oops", err)
}
```

This might seem pointless, but can be useful if the error is handled further up
the call stack. The standard [net package uses this style](https://godoc.org/net#Error) for special conditions.

We can even get the easy syntax back with a helper function:

```go
head, err := client.BlockByHash(ctx, common.Hash{})
if err != nil && !ethereum.IsNotFound(err) {
     ... 
}
// head might still be nil, but the check is explicit.
```

### What should it be then?

Please provide feedback and tell me about your preferred choice.

### PR DESCRIPTION
**Pull request by fjl**
This PR contains API changes split out of #3250 and adds an error for missing items. It also moves the Removed Log field from eth/filters to core/vm so it can be unmarshaled from RPC responses when using ethclient.

See commit messages for more detail about the changes.

Closes #3379
Fixes #3331

### CODE REVIEW

File: ethclient/ethclient_test.go [hunk 1]
```diff
@@ -21,9 +21,9 @@ import "github.com/ethereum/go-ethereum"
 // Verify that Client implements the ethereum interfaces.
 var (
 	_ = ethereum.ChainReader(&Client{})
+	_ = ethereum.TransactionReader(&Client{})
```
Comment (bas-vk) [inline ethclient/ethclient_test.go]: The client currently doesn't satisfy the interface:
```
# github.com/ethereum/go-ethereum/ethclient
./ethclient_test.go:24: cannot convert Client literal (type *Client) to type ethereum.TransactionReader:
	*Client does not implement ethereum.TransactionReader (wrong type for TransactionByHash method)
		have TransactionByHash("golang.org/x/net/context".Context, common.Hash) (*types.Transaction, error)
		want TransactionByHash("golang.org/x/net/context".Context, common.Hash) (*types.Transaction, bool, error)
FAIL	github.com/ethereum/go-ethereum/ethclient [build failed]
```

Comment (fjl) [inline ethclient/ethclient_test.go]: oops, I forgot to add this change ;)

Comment (fjl) [inline ethclient/ethclient_test.go]: PTAL

Comment (fjl) [inline ethclient/ethclient_test.go]: ahh, wait, that won't work

File: ethclient/ethclient.go [hunk 2]
```diff
@@ -155,17 +155,20 @@ func (ec *Client) HeaderByNumber(ctx context.Context, number *big.Int) (*types.H
...
-	if err == nil {
-		if tx == nil {
-			err = ethereum.NotFound
-		} else if _, r, _ := tx.RawSignatureValues(); r == nil {
-			return nil, fmt.Errorf("server returned transaction without signature")
-		}
+func (ec *Client) TransactionByHash(ctx context.Context, hash common.Hash) (tx *types.Transaction, isPending bool, err error) {
+	var rtx *struct {
+		*types.Transaction
+		BlockHash *common.Hash
```
Comment (bas-vk) [inline ethclient/ethclient.go]: The response also contains a block number that can be interesting for users.

**Comment by bas-vk** [comment]
LGTM 👍

**Comment by fjl** [comment]
Squashed. @bas-vk did you test this somehow?

**Comment by zsfelfoldi** [comment]
LGTM :+1:

**Pull request merged**